### PR TITLE
[analysis] Add trace preprocessing

### PR DESCRIPTION
**Pull request by nasahlpa**
This PR adds a trace preprocessing script allowing the user to (a) filter traces using a max. amount of tolerable deviation from the trace average and (b) align the traces using a reference window. Currently, the script is tested with the new trace database format. Moreover, I've decided to implement this functionality outside of the TVLA script as this script is already really overblown.

Input trace set:
image
After filtering:
image
After aligning:
image
The aligning (which uses functionality from ChipWhisperer) heavily depends on the selected alignment window.

### CODE REVIEW

File: analysis/trace_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    else:
+        return project_in
+
+
+def align_traces(args: dict, project_in: SCAProject, project_out: SCAProject):
+    """ Align traces.
+
+    This function filters all traces that contain values over the tolerable
+    deviation from average (num_sigmas). Only up to max_traces_mem traces are
+    held in memory.
```
Comment (vrozic) [inline analysis/trace_preprocessing.py:206]: It looks like this description got accidentally copied from `filter_traces()`

File: analysis/trace_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+                        "--align_enable",
+                        dest="align_enable",
+                        action='store_true',
+                        help="Aligtn the traces")
+    parser.add_argument("-s",
+                        "--num_sigmas",
+                        dest="num_sigmas",
+                        type=float,
+                        required=False,
+                        default=100,
```
[the remark this answers is not in the document]
Comment (nasahlpa) [reply analysis/trace_preprocessing.py]: As this value depends on the trace set, I've now deleted it as we really should force the user to define this value.

**Review by vrozic — APPROVED** [review]
Thanks @nasahlpa for the fixes. The code looks good to me. Nice work.

If I interpret the pictures correctly, `filter_traces()` will get rid of the traces that are heavily misaligned; afterwards `align_traces()` will horizontally shift traces that are slightly misaligned.

Could you please add a paragraph to the getting started guide with instructions on how to run the pre-processing and what it achieves?

**Comment by nasahlpa** [comment]
Done. Thanks for the review @vrozic!

**Pull request merged**